### ISSUE TITLE
Generating a cluster state archive tool: fatal error: runtime: out of memory

### ISSUE THREAD

**Issue by itsvikasgupta** [istio/istio#12510]
I am seeing an error:

"fatal error: runtime: out of memory"

while dumping cluster state.

     $ curl https://raw.githubusercontent.com/istio/istio/release-1.0/tools/dump_kubernetes.sh | sh -s -- -z
      % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                     Dload  Upload   Total   Spent    Left  Speed
    100  7837  100  7837    0     0  38776      0 --:--:-- --:--:-- --:--:-- 38990
    Fetching debug/configz from pilot
    Unable to use a TTY - input is not a terminal or the right kind of file
      % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                     Dload  Upload   Total   Spent    Left  Speed
    100 10577    0 10577    0     0  1445k      0 --:--:-- --:--:-- --:--:-- 1475k
    Fetching debug/endpointz from pilot
    Unable to use a TTY - input is not a terminal or the right kind of file
      % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                     Dload  Upload   Total   Spent    Left  Speed
    100 75797    0 75797    0     0  9147k      0 --:--:-- --:--:-- --:--:-- 10.3M
    Fetching debug/adsz from pilot
    Unable to use a TTY - input is not a terminal or the right kind of file
      % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                     Dload  Upload   Total   Spent    Left  Speed
    100  1069  100  1069    0     0   212k      0 --:--:-- --:--:-- --:--:--  260k
    Fetching metrics from pilot
    Unable to use a TTY - input is not a terminal or the right kind of file
      % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                     Dload  Upload   Total   Spent    Left  Speed
    100 16767    0 16767    0     0  2637k      0 --:--:-- --:--:-- --:--:-- 2729k
    Retrieving kubernetes resource configurations
    Retrieving istio resource configurations

    fatal error: runtime: out of memory

    runtime stack:
    runtime.throw(0x17d2253, 0x16)
            /usr/local/go/src/runtime/panic.go:608 +0x72
    runtime.sysMap(0xc044000000, 0x40000000, 0x29969b8)
            /usr/local/go/src/runtime/mem_linux.go:156 +0xc7
    runtime.(*mheap).sysAlloc(0x297d180, 0x40000000, 0x0, 0x0)
            /usr/local/go/src/runtime/malloc.go:619 +0x1c7
    runtime.(*mheap).grow(0x297d180, 0x20000, 0x0)
            /usr/local/go/src/runtime/mheap.go:920 +0x42
    runtime.(*mheap).allocSpanLocked(0x297d180, 0x20000, 0x29969c8, 0x43310d)
            /usr/local/go/src/runtime/mheap.go:848 +0x337
    runtime.(*mheap).alloc_m(0x297d180, 0x20000, 0xffffffffffff0101, 0x7ffe2cd1b120)
            /usr/local/go/src/runtime/mheap.go:692 +0x119
    runtime.(*mheap).alloc.func1()
            /usr/local/go/src/runtime/mheap.go:759 +0x4c
    runtime.(*mheap).alloc(0x297d180, 0x20000, 0x7ffe2c010101, 0x7ffe2cd1b1a0)
            /usr/local/go/src/runtime/mheap.go:758 +0x8a
    runtime.largeAlloc(0x3ffffe00, 0x7f9aeef20101, 0x1)
            /usr/local/go/src/runtime/malloc.go:1019 +0x97
    runtime.mallocgc.func1()
            /usr/local/go/src/runtime/malloc.go:914 +0x46
    runtime.systemstack(0x4582b9)
            /usr/local/go/src/runtime/asm_amd64.s:351 +0x66
    runtime.mstart()
            /usr/local/go/src/runtime/proc.go:1229

    goroutine 1 [running]:
    runtime.systemstack_switch()
            /usr/local/go/src/runtime/asm_amd64.s:311 fp=0xc00046d278 sp=0xc00046d270 pc=0x4583b0
    runtime.mallocgc(0x3ffffe00, 0x1575d40, 0x1, 0xc00046d350)
            /usr/local/go/src/runtime/malloc.go:913 +0x896 fp=0xc00046d318 sp=0xc00046d278 pc=0x40c136
    runtime.makeslice(0x1575d40, 0x3ffffe00, 0x3ffffe00, 0xc0000f6e50, 0x0, 0x0)
            /usr/local/go/src/runtime/slice.go:70 +0x77 fp=0xc00046d348 sp=0xc00046d318 pc=0x443067
    bytes.makeSlice(0x3ffffe00, 0x0, 0x0, 0x0)
            /usr/local/go/src/bytes/buffer.go:231 +0x6d fp=0xc00046d388 sp=0xc00046d348 pc=0x4eb17d
    bytes.(*Buffer).grow(0xc0008b8070, 0x200, 0xe00)
            /usr/local/go/src/bytes/buffer.go:144 +0x16c fp=0xc00046d3d8 sp=0xc00046d388 pc=0x4eaadc
    bytes.(*Buffer).ReadFrom(0xc0008b8070, 0x7f9aeef55560, 0xc0007a6b40, 0xc00002e000, 0x1610a60, 0x1672b40)
            /usr/local/go/src/bytes/buffer.go:204 +0x4b fp=0xc00046d448 sp=0xc00046d3d8 pc=0x4eaf6b
    io/ioutil.readAll(0x7f9aeef55560, 0xc0007a6b40, 0x200, 0x0, 0x0, 0x0, 0x0, 0x0)
            /usr/local/go/src/io/ioutil/ioutil.go:36 +0xb5 fp=0xc00046d490 sp=0xc00046d448 pc=0x5d56a5
    io/ioutil.ReadAll(0x7f9aeef55560, 0xc0007a6b40, 0xc0007a6b40, 0x7f9aeef55560, 0xc0007a6b40, 0xc0008c3538, 0x42ad12)
            /usr/local/go/src/io/ioutil/ioutil.go:45 +0x3e fp=0xc00046d4e0 sp=0xc00046d490 pc=0x5d576e
    k8s.io/kubernetes/vendor/k8s.io/client-go/rest.(*Request).DoRaw.func1(0xc000434700, 0xc000916a20)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/rest/request.go:819 +0x72 fp=0xc00046d548 sp=0xc00046d4e0 pc=0xecc172
    k8s.io/kubernetes/vendor/k8s.io/client-go/rest.(*Request).request.func2(0xc000916a20, 0xc0008c3648, 0xa, 0xc000740f00, 0xc0008c3720, 0xc000434700, 0xc0001021c0, 0x7d, 0x0)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/rest/request.go:783 +0xca fp=0xc00046d5f0 sp=0xc00046d548 pc=0xecbd0a
    k8s.io/kubernetes/vendor/k8s.io/client-go/rest.(*Request).request(0xc000740f00, 0xc00046d720, 0x0, 0x0)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/rest/request.go:785 +0x6ac fp=0xc00046d700 sp=0xc00046d5f0 pc=0xec71dc
    k8s.io/kubernetes/vendor/k8s.io/client-go/rest.(*Request).DoRaw(0xc000740f00, 0xc00033baa0, 0x19b18e0, 0xc000a64000, 0x19b19e0, 0xc000bb2d80)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/rest/request.go:818 +0xa3 fp=0xc00046d798 sp=0xc00046d700 pc=0xec7993
    k8s.io/kubernetes/pkg/kubectl/cmd/clusterinfo.(*ClusterInfoDumpOptions).Run.func1(0x19a9a00, 0xc00000c020, 0xc00064abc0, 0xb, 0xc0005a6f60, 0x1a, 0x0, 0x0, 0x0, 0x0, ...)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/clusterinfo/clusterinfo_dump.go:267 +0x407 fp=0xc00046d8e0 sp=0xc00046d798 pc=0x11bc457
    k8s.io/kubernetes/pkg/kubectl/cmd/clusterinfo.(*ClusterInfoDumpOptions).Run(0xc00095c600, 0x0, 0xc00033baa0)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/clusterinfo/clusterinfo_dump.go:283 +0x240 fp=0xc00046dc78 sp=0xc00046d8e0 pc=0x11ba4b0
    k8s.io/kubernetes/pkg/kubectl/cmd/clusterinfo.NewCmdClusterInfoDump.func1(0xc000b8f400, 0x2994bf0, 0x0, 0x0)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/clusterinfo/clusterinfo_dump.go:79 +0x75 fp=0xc00046dcc0 sp=0xc00046dc78 pc=0x11bc015
    k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute(0xc000b8f400, 0x2994bf0, 0x0, 0x0, 0xc000b8f400, 0x2994bf0)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:760 +0x2cc fp=0xc00046ddb0 sp=0xc00046dcc0 pc=0x66a36c
    k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc000195180, 0xc00024d9e0, 0x12a05f200, 0xc00088bf28)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:846 +0x2fd fp=0xc00046def0 sp=0xc00046ddb0 pc=0x66af0d
    k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute(0xc000195180, 0x187bed0, 0x2976160)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:794 +0x2b fp=0xc00046df20 sp=0xc00046def0 pc=0x66abeb
    main.main()
            _output/dockerized/go/src/k8s.io/kubernetes/cmd/kubectl/kubectl.go:50 +0xf3 fp=0xc00046df98 sp=0xc00046df20 pc=0x1483603
    runtime.main()
            /usr/local/go/src/runtime/proc.go:201 +0x207 fp=0xc00046dfe0 sp=0xc00046df98 pc=0x42d927
    runtime.goexit()
            /usr/local/go/src/runtime/asm_amd64.s:1333 +0x1 fp=0xc00046dfe8 sp=0xc00046dfe0 pc=0x45a311

    goroutine 5 [chan receive]:
    k8s.io/kubernetes/vendor/k8s.io/klog.(*loggingT).flushDaemon(0x29764e0)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/klog.go:941 +0x8b
    created by k8s.io/kubernetes/vendor/k8s.io/klog.init.0
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/klog.go:403 +0x69

    goroutine 6 [syscall]:
    os/signal.signal_recv(0x0)
            /usr/local/go/src/runtime/sigqueue.go:139 +0x9c
    os/signal.loop()
            /usr/local/go/src/os/signal/signal_unix.go:23 +0x22
    created by os/signal.init.0
            /usr/local/go/src/os/signal/signal_unix.go:29 +0x41

    goroutine 9 [select]:
    k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x187de18, 0x12a05f200, 0x0, 0x1, 0xc00004afc0)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:145 +0x147
    k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0x187de18, 0x12a05f200, 0xc00004afc0)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
    created by k8s.io/kubernetes/pkg/kubectl/util/logs.InitLogs
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/util/logs/logs.go:51 +0x89

    goroutine 17 [IO wait]:
    internal/poll.runtime_pollWait(0x7f9aeef23f00, 0x72, 0xc000456870)
            /usr/local/go/src/runtime/netpoll.go:173 +0x66
    internal/poll.(*pollDesc).wait(0xc00074c718, 0x72, 0xffffffffffffff00, 0x19ab480, 0x27c9578)
            /usr/local/go/src/internal/poll/fd_poll_runtime.go:85 +0x9a
    internal/poll.(*pollDesc).waitRead(0xc00074c718, 0xc0000e0000, 0x8000, 0x8000)
            /usr/local/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
    internal/poll.(*FD).Read(0xc00074c700, 0xc0000e0000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
            /usr/local/go/src/internal/poll/fd_unix.go:169 +0x179
    net.(*netFD).Read(0xc00074c700, 0xc0000e0000, 0x8000, 0x8000, 0x408a7b, 0xc00002e000, 0x1610a60)
            /usr/local/go/src/net/fd_unix.go:202 +0x4f
    net.(*conn).Read(0xc00000d130, 0xc0000e0000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
            /usr/local/go/src/net/net.go:177 +0x68
    crypto/tls.(*block).readFromUntil(0xc000714600, 0x19a95c0, 0xc00000d130, 0x5, 0xc00000d130, 0x0)
            /usr/local/go/src/crypto/tls/conn.go:492 +0x89
    crypto/tls.(*Conn).readRecord(0xc00061a700, 0x187e817, 0xc00061a820, 0x187e848)
            /usr/local/go/src/crypto/tls/conn.go:593 +0xdd
    crypto/tls.(*Conn).Read(0xc00061a700, 0xc000710000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
            /usr/local/go/src/crypto/tls/conn.go:1145 +0xf1
    bufio.(*Reader).Read(0xc00007b560, 0xc00015b378, 0x9, 0x9, 0x7fb8b0, 0x42de3c, 0xc000456c50)
            /usr/local/go/src/bufio/bufio.go:216 +0x22f
    io.ReadAtLeast(0x19a6ca0, 0xc00007b560, 0xc00015b378, 0x9, 0x9, 0x9, 0x187e848, 0xc000456cf0, 0xc000456ce0)
            /usr/local/go/src/io/io.go:310 +0x88
    io.ReadFull(0x19a6ca0, 0xc00007b560, 0xc00015b378, 0x9, 0x9, 0xc00080a000, 0x1000, 0x4000)
            /usr/local/go/src/io/io.go:329 +0x58
    k8s.io/kubernetes/vendor/golang.org/x/net/http2.readFrameHeader(0xc00015b378, 0x9, 0x9, 0x19a6ca0, 0xc00007b560, 0x0, 0x0, 0x0, 0x0)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/frame.go:237 +0x7b
    k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*Framer).ReadFrame(0xc00015b340, 0xc040ce8060, 0x0, 0x0, 0x0)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/frame.go:492 +0xa3
    k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*clientConnReadLoop).run(0xc000456fb8, 0x187d280, 0xc00005a7b8)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/transport.go:1477 +0x9e
    k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*ClientConn).readLoop(0xc0000f6e00)
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/transport.go:1405 +0x76
    created by k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*Transport).newClientConn
            /workspace/anago-v1.13.3-beta.0.37+721bfa751924da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/transport.go:592 +0x60d
    Retrieving logs for copy_core_dumps_if_istio_proxy//
    error: You must provide one or more resources by argument or filename.
    Example resource specifications include:
       '-f rsrc.yaml'
       '--filename=rsrc.json'
       '<resource> <name>'
       '<resource>'
    error: resource name may not be empty
    sh: line 108: [: : integer expression expected
    Wrote ./istio-dump.tar.gz
    Wrote to istio-dump
    sh: line 283: return: tap_containers: numeric argument required

**Version**
    $ istioctl version
    
    Version: 1.0.5
    GitRevision: c1707e45e71c75d74bf3a5dec8c7086f32f32fad
    User: root@6f6ea1061f2b
    Hub: docker.io/istio
    GolangVersion: go1.10.4
    BuildStatus: Clean

    $ kubectl version

    Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.3", GitCommit:"721bfa751924da8d1680787490c54b9179b1fed0", GitTreeState:"clean", BuildDate:"2019-02-01T20:08:12Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
    Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.4", GitCommit:"5ca598b4ba5abb89bb773071ce452e33fb66339d", GitTreeState:"clean", BuildDate:"2018-06-06T08:00:59Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

**Environment**

    AWS
    Red Hat Enterprise Linux Server release 7.6 (Maipo)
    GNU bash, version 4.2.46(2)-release (x86_64-redhat-linux-gnu)

**Installation**
I used the istio-demo-auth.yaml file in the istio-1.0.5 release and did:

        $ kubectl apply -f istio-demo-auth.yaml

**Comment by esnible** [comment]
I am not sure why you are seeing the Out Of Memory error.  I do know that fixes were made a few months ago to make sure that gigantic log files are not included.  I am not sure which versions have that fix.

The dump_kubenetes.sh tool has some problems running on bash 3 (Mac OSX) but not on bash 4 (Linux).  The work item is https://github.com/istio/istio/issues/12156 and there is a PR https://github.com/istio/istio/pull/12159 which I am hoping to get into 1.1.1 but not the upcoming 1.1.

**Comment by vikasguptaanaplan** [comment]
I ran  
    
    ./dump_kubernetes.sh with bash -x

Here is where the OOM happens:

    + kubectl cluster-info dump
    fatal error: runtime: out of memory

**Comment by vikasguptaanaplan** [comment]
I ran

    $ kubectl --v=9 cluster-info dump

Looks like it is OOMing while getting the logs of calico-node container. I wonder if there is a way to tell cluster-info to skip that container.

    ==== START logs for container calico-node of pod kube-system/calico-node-2qksq ====
    I0315 17:26:05.286403    8198 loader.go:359] Config loaded from file /home/company/.kube/config
    I0315 17:26:05.286750    8198 round_trippers.go:419] curl -k -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: kubectl/v1.13.3 (linux/amd64) kubernetes/721bfa7" 'https://istiodev-haproxy-api-1.company.io:6443/api/v1/namespaces/kube-system/pods/calico-node-2qksq/log?container=calico-node'
    I0315 17:26:05.308920    8198 round_trippers.go:438] GET https://istiodev-haproxy-api-1.company.io:6443/api/v1/namespaces/kube-system/pods/calico-node-2qksq/log?container=calico-node 200 OK in 22 milliseconds
    I0315 17:26:05.308942    8198 round_trippers.go:444] Response Headers:
    I0315 17:26:05.308949    8198 round_trippers.go:447]     Content-Type: text/plain
    I0315 17:26:05.308953    8198 round_trippers.go:447]     Date: Fri, 15 Mar 2019 17:26:05 GMT

    fatal error: runtime: out of memory

**Comment by esnible** [comment]
I checked the source code for `cluster-info dump` (at https://github.com/kubernetes/kubernetes/blob/f229aa0ad65a0378645a68ebc7ca2bf65e4957e9/pkg/kubectl/cmd/clusterinfo/clusterinfo_dump.go ); no options other than restricting namespaces.

I ran with and without `-v 9`.  For me, the command takes 90 seconds and produces a 153 MB file.  I am using _kubectl_ v1.13.4.  I found nothing in github about people getting OOM with this command.
